src/Storages/MergeTree/ReplicatedMergeTreeSink.cpp

```diff
@@ -197,9 +335,21 @@ void ReplicatedMergeTreeSink::consume(Chunk chunk)
     if (!storage_snapshot->object_columns.empty())
         convertDynamicColumnsToTuples(block, storage_snapshot);
 
-    auto part_blocks = storage.writer.splitBlockIntoParts(block, max_parts_per_block, metadata_snapshot, context);
 
-    using DelayedPartitions = std::vector<ReplicatedMergeTreeSink::DelayedChunk::Partition>;
+    ChunkOffsetsPtr chunk_offsets;
+
+    if constexpr (async_insert)
+    {
+        const auto & chunk_info = chunk.getChunkInfo();
+        if (const auto * chunk_offsets_ptr = typeid_cast<const ChunkOffsets *>(chunk_info.get()))
+            chunk_offsets = std::make_shared<ChunkOffsets>(chunk_offsets_ptr->offsets);
+        else
+            throw Exception("No chunk info for async inserts", ErrorCodes::LOGICAL_ERROR);
+    }
+
+    auto part_blocks = storage.writer.splitBlockIntoParts(block, max_parts_per_block, metadata_snapshot, context, chunk_offsets);
+
+    using DelayedPartitions = std::vector<typename ReplicatedMergeTreeSink<async_insert>::DelayedChunk::Partition>;
```

Comment:
Better to make some alias for `ReplicatedMergeTreeSink<async_insert>::DelayedChunk::Partition` if possible.